netmap/netmap_contract.go

```diff
@@ -601,6 +608,37 @@ func ListConfig() []record {
 	return config
 }
 
+// SubscribeForNewEpoch registers passed contract as a NewEpoch event
+// subscriber. Such a contract must have a `NewEpoch` method with a single
+// numeric parameter. Transactions that call SubscribeForNewEpoch must be
+// witnessed by the Alphabet.
+// Produces `NewEpochSubscription` notification event with a just
+// registered recipient in a success case.
+func SubscribeForNewEpoch(contract interop.Hash160) {
+	common.CheckAlphabetWitness(common.AlphabetAddress())
+
+	callingContract := runtime.GetExecutingScriptHash()
+	if !management.HasMethod(callingContract, "newEpoch", 1) {
+		panic(address.FromHash160(callingContract) + " contract does not have `newEpoch(epoch)` method")
+	}
+
+	ctx := storage.GetContext()
+	var num byte
+	it := storage.Find(ctx, []byte(newEpochSubscribersPrefix), storage.None)
+	for iterator.Next(it) {
+		num += 1
+	}
+
+	var key []byte
+	key = append(key, []byte(newEpochSubscribersPrefix)...)
+	key = append(key, num)
+	key = append(key, contract...)
+
+	storage.Put(ctx, key, []byte{})
```

Comment:
@roman-khimov, wrt https://github.com/nspcc-dev/neofs-contract/pull/368#discussion_r1403628376 and https://github.com/nspcc-dev/neofs-contract/pull/368#discussion_r1402727722: we need ordering guarantees and it's good. But this scheme currently allows subscriptions duplicates I think that we need to disallow subscriptions duplicates on the netmap contract level, and in this part of code we need to add a check whether subscription is already added.

What do you think?

Reply:
done with ignoring (no logs or panics)